### PR TITLE
XW-297 Refactor and optimize all verbatim tag replacements

### PR DESCRIPTION
**Pull request by rogatty**
https://wikia-inc.atlassian.net/browse/XW-297

**This is work in progress.**

/cc @nandy-andy @Warkot @bognix @vforge

### CODE REVIEW

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 1]
```diff
@@ -40,28 +40,30 @@ class TwitterTagController extends WikiaController {
...
-		if ( empty( $args['widget-id'] ) ) {
-			return '<strong class="error">' . wfMessage( 'twitter-tag-widget-id' )->parse() . '</strong>';
-		}
+	protected function getAttributesAllowed() {
+		return array_keys( self::TAG_PERMITTED_ATTRIBUTES );
+	}
+
+	protected function getErrorOutput( $errorMessages ) {
+		return $this->sendRequest(
+			'TwitterTagController',
```
Comment (vforge) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php:53]: You can use `sendSelfRequest`

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 2]
```diff
@@ -69,37 +71,42 @@ public function parseTag( $input, array $args, Parser $parser, PPFrame $frame )
...
-
-		return $validatedAttributes;
 	}
 
 	protected function buildParamValidator( $paramName ) {
-		return true;
+		$validator = new WikiaValidatorAlwaysTrue();
+		$attributesRegexPatterns = self::TAG_PERMITTED_ATTRIBUTES;
+
+		switch ( $paramName ) {
```
Comment (vforge) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php:89]: Have you considered adding validators for all parameters?

File: extensions/wikia/TwitterTag/TwitterTagController.class.php (hunk 1 above)
Comment (nandy-andy) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php:53]: Good point! I've used this method so rarely I didn't think of that either. Thanks V.! :smile_cat:

File: extensions/wikia/TwitterTag/TwitterTagController.class.php [hunk 3]
```diff
@@ -40,28 +40,30 @@ class TwitterTagController extends WikiaController {
...
 
-		$attributes = $this->prepareAttributes( $args, self::TAG_PERMITTED_ATTRIBUTES );
+	protected function getSuccessOutput( $args ) {
+		$attributes = $this->buildTagAttributes( self::TAG_PERMITTED_ATTRIBUTES, $args, 'data' );
 		$attributes['href'] = self::TWITTER_BASE_URL;
+
 		// data-wikia-widget attribute is searched for by Mercury
 		$attributes['data-wikia-widget'] = self::PARSER_TAG_NAME;
 
-		if ( ( new WikiaIFrameTagBuilderHelper() )->isMobileSkin() ) {
```
Comment (nandy-andy) [inline extensions/wikia/TwitterTag/TwitterTagController.class.php:64]: Oh, we'll have to sync it somehow with the Flite tag/widget @fraszczakszymon was working on recently: https://github.com/Wikia/app/pull/8806/files#diff-08ba9d6c112d52c6530b470657257f75R24

**Pull request closed without merging**